### ISSUE TITLE
Add ability to use own proxy in private instance

### ISSUE THREAD

**Issue by krasnoludkolo** [Benjamin-Loison/YouTube-operational-API#249]
Basically like in title. I would like to be able to route traffic from my private instance via my proxies.

**Comment by Benjamin-Loison** [comment]
What is the point of using proxies in your case? Are you facing #11 or you are looking for faster bulk process? Furthermore what kind of proxies (HTTPS, SOCKS, ...) on what OS (if Linux precise which distribution) you have in mind?

**Comment by krasnoludkolo** [comment]
I want to be able to scale up with my private instance and I don't want to put my IP addresses at risk of being blocked by YT

**Comment by krasnoludkolo** [comment]
> what kind of proxies (HTTPS, SOCKS, ...)

 http/https
 
 > on what OS (if Linux precise which distribution) you have in mind? 

 <details>
<summary>Does it matter? I would like to achieve something similar to python</summary>
 
```py
 import requests as r
 proxies = {
    'http': proxy_url,
    'https': proxy_url,
}
 
r.get(
    url,
    proxies=proxies
)
```
</details>
 
 > Just to make sure, you would like to use a single or multiple proxies at the same time?

 single proxy configuration.
 
 
 This could be configured via env variables of private instance for example

**Comment by Benjamin-Loison** [comment]
I asked concerning the OS, as I was not sure about your use case and depending on the use case I could redirect you to a given OS proxy feature but this does not seem accurate in your case.

I am not keen with HTTP(S) proxies, I will investigate and try to solve your issue this evening.

For a similar purpose of not being detected as abusing with YouTube servers, [I used Tor to retrieve many captions of videos](https://codeberg.org/Benjamin_Loison/yt_100_most_subscribed_fr/src/commit/2af592cb4ff790eab47ac456ff8ccbd3ee39a222/generation/captions/captions/extract_captions.py) and it worked very nicely but that was in Python and not using YouTube operational API.

I was aware how to use SOCKS proxy with `ssh -ND 9999 lemnoslife.com` and [Tor](https://codeberg.org/Benjamin_Loison/yt_100_most_subscribed_fr/src/commit/2af592cb4ff790eab47ac456ff8ccbd3ee39a222/generation/captions/captions/extract_captions.py#L35).

[Wikipedia: Proxy server#Web proxy servers (1212429758)](https://en.wikipedia.org/w/index.php?title=Proxy_server&oldid=1212429758#Web_proxy_servers)

[Wikipedia: HTTP tunnel#HTTP CONNECT method (1184773834)](https://en.wikipedia.org/w/index.php?title=HTTP_tunnel&oldid=1184773834#HTTP_CONNECT_method)

Asked on [Discord](https://discord.com/channels/@me/1210337227876597770/1215445417635090482) how to setup a server HTTPS proxy.

```bash
curl -4 https://ifconfig.me
```
```
curl: (56) Proxy CONNECT aborted
```

while:

```bash
a2enmod proxy
```

<details>
<summary>and added to <code>apache2.conf</code>:</summary>

```
ProxyRequests On
ProxyVia On

<Proxy "*">
  Require ip X.X.X.X
</Proxy>
```
</details>

and of course:

```bash
sudo service apache2 restart
```

https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#examples
https://httpd.apache.org/docs/2.4/mod/mod_proxy.html#access

**Comment by Benjamin-Loison** [comment]
<details>
<summary>Bash script:</summary>

```bash
export http_proxy=https://CENSORED.yt.lemnoslife.com:443/; curl -4 http://ifconfig.me
export https_proxy=https://CENSORED.yt.lemnoslife.com:443/; curl -4 https://ifconfig.me
```
</details>

work as expected. With another IP than the authorized one I get:

```
curl: (56) CONNECT tunnel failed, response 403
```

`curl: (56) CONNECT tunnel failed, response 403` does not return interesting results with DuckDuckGo and Google.

<details>
<summary>On server side I did:</summary>

```bash
a2enmod proxy_connect
a2enmod proxy
a2enmod proxy_http
service apache2 restart
```
</details>

<details>
<summary><code>apache2.conf</code>:</summary>

```
ProxyRequests On
ProxyVia On

<Proxy "*">
  Require ip X.X.X.X
  Require ip X:X:X:X:X:X:X:X
</Proxy>
```
</details>

Note that everything is maybe not necessary.

The aim is to avoid using cURL otherwise it would require a lot of refactoring.

Forcing PHP to use IPv4 thanks to [Benjamin-Loison/cinnamon/issues/16](https://github.com/Benjamin-Loison/cinnamon/issues/16).

Can use `52.211.130.173` (AWS) instead of `ifconfig.me` to avoid possible domain resolution issues. Using explicit IPv4 of `CENSORED.yt.lemnoslife.com` does not help too in:

<details>
<summary>PHP script:</summary>

```php
$context = stream_context_create([
    'http' => [
        'proxy' => 'tcp://X.X.X.X:443',
        'request_fulluri' => true,
    ]
]);

echo file_get_contents('http://52.211.130.173', false, $context);
```
</details>

As I get:

```
PHP Warning:  file_get_contents(https://ifconfig.me): Failed to open stream: Cannot connect to HTTPS server through proxy in /home/benjamin/test.php on line 10
```

With `80` instead of `443` it works.

<details>
<summary>PHP script:</summary>

```php
$context = stream_context_create([
    'http' => [
        'proxy' => 'tcp://CENSORED.yt.lemnoslife.com:80',
        'request_fulluri' => true,
    ]
]);

echo file_get_contents('https://ifconfig.me', false, $context);
```
</details>

works.

Without authentication with login and username, it seems that above `80` port usage is secure as with Wireshark I am unable to read packets while when using `http://ifconfig.me` I am able to.

<details>
<summary>Note that if use <code>ssl</code> instead of <code>tcp</code> and <code>443</code> instead of <code>80</code>, then I get:</summary>

```
PHP Warning:  file_get_contents(): SSL/TLS already set-up for this stream in /home/benjamin/test.php on line 10
PHP Warning:  file_get_contents(https://ifconfig.me): Failed to open stream: Cannot connect to HTTPS server through proxy in /home/benjamin/test.php on line 10
```
</details>

While used https://www.php.net/manual/en/context.http.php, https://www.php.net/manual/en/context.ssl.php does not seem relevant.
https://www.php.net/manual/en/function.stream-context-create.php

Note that YouTube operational API is only HTTPS based, so there is no need to test HTTP proxy, only HTTPS is necessary.

Concerning authentication:

Trying [the Stack Overflow answer 724656](https://stackoverflow.com/a/724656).

Now getting:

```bash
export https_proxy=https://CENSORED.yt.lemnoslife.com:443/; curl -4 https://ifconfig.me
```
```
curl: (56) CONNECT tunnel failed, response 407
```

```bash
export https_proxy=https://username:CORRECT_PASSWORD@CENSORED.yt.lemnoslife.com:443/; curl -4 https://ifconfig.me
```
```
X.X.X.X
```

```bash
export https_proxy=https://username:INCORRECT_PASSWORD@CENSORED.yt.lemnoslife.com:443/; curl -4 https://ifconfig.me
```
```
curl: (56) CONNECT tunnel failed, response 407
```

Now need to a PHP version of that.

With above PHP I get:

```bash
php test.php
```
```
PHP Warning:  file_get_contents(https://ifconfig.me): Failed to open stream: Cannot connect to HTTPS server through proxy in /home/benjamin/test.php on line 10
```

<details>
<summary>PHP script:</summary>

```php
$auth = base64_encode('username:CORRECT_PASSWORD');

$context = stream_context_create([
    'http' => [
        'proxy' => 'tcp://CENSORED.yt.lemnoslife.com:80',
        'request_fulluri' => true,
        'header' => "Proxy-Authorization: Basic $auth",
    ]
]);

echo file_get_contents('https://ifconfig.me', false, $context);
```
</details>

```
X.X.X.X
```

if does not type the correct password, then get:

```
PHP Warning:  file_get_contents(https://ifconfig.me): Failed to open stream: Cannot connect to HTTPS server through proxy in /home/benjamin/test.php on line 13
```

So ready to implement this feature.

```bash
grep -rw 'file_get_contents' | grep -vE "file_get_contents\(('tests/|'.git/|KEYS_FILE)"
```

Only interesting occurrence:

https://github.com/Benjamin-Loison/YouTube-operational-API/blob/eaec310cc064872ca6fff65712315b91b7fa2e65/common.php#L67

Maybe ask @krasnoludkolo if my commit meets his needs before pushing/deploying.
Making a pull request seems to make sense.

```bash
grep -r 'getRemote('
```

Making add key not use HTTPS proxy would be preferable but as it is a rare event, let us postpone doing so.

It seems that `noKey/index.php` uses its own `file_get_contents` for YouTube Data API v3 retrieval, so can assume that all requests to `getRemote` (more precisely `fileGetContentsAndHeadersFromOpts`) can use HTTPS proxy, if any is configured.

```bash
grep -rvE "c4TabbedHeaderRenderer|charset|\['header'\] = \['|'header' => \['" --include=*.php | grep -w 'header'
```

<details>
<summary>Just a reminder I set <code>HTTPS_PROXY_PORT</code> to <code>80</code>:</summary>

```diff
+        $url = 'https://checkip.amazonaws.com';
         $result = file_get_contents($url, false, $context);
+        die($result);
```
</details>

```bash
curl -s 'http://localhost/YouTube-operational-API/search?part=snippet&q=test'
```

With server configured for HTTPS proxy with authentication required get an empty output and get in `/var/log/apache2/error.log`:
```
[php:warn] [pid 1587] [client 127.0.0.1:53520] PHP Warning:  file_get_contents(https://checkip.amazonaws.com): Failed to open stream: Cannot connect to HTTPS server through proxy in /var/www/html/YouTube-operational-API/common.php on line 83
```

If I provide the correct credentials in `configuration.php`, I correctly get the proxy IP.

If I switch to IP verification, I correctly get the proxy IP but if I switch the IP verification to another IP, then I correctly have an empty output, but have these logs:

```
[php:warn] [pid 1590] [client 127.0.0.1:35612] PHP Warning:  file_get_contents(https://checkip.amazonaws.com): Failed to open stream: Cannot connect to HTTPS server through proxy in /var/www/html/YouTube-operational-API/common.php on line 83
```

Disabling HTTPS proxy usage in `configuration.php` shows the correct actual IP.

**Comment by Benjamin-Loison** [comment]
@krasnoludkolo Does the proxy you have in mind requires authentication requiring a login and password? If so I will consider [the Stack Overflow answer 1336419](https://stackoverflow.com/a/1336419).

**Comment by krasnoludkolo** [comment]
I use format:
```
http://username:password:hostname:port
```

**Comment by krasnoludkolo** [comment]
So in other places tools allows to pass `proxy_url` as parameter and I just use this to pass my proxy url with auth info within

**Comment by Benjamin-Loison** [comment]
@krasnoludkolo May you confirm me that the following `diff` based on 48f8e2d3a7659700ce58b2ccc5357893100e3736 solves your issue?

<details>
<summary><code>diff</code>:</summary>

```diff
diff --git a/common.php b/common.php
index 0f4585f..5ce9a23 100644
--- a/common.php
+++ b/common.php
@@ -63,8 +63,25 @@
 
     function fileGetContentsAndHeadersFromOpts($url, $opts)
     {
+        if(HTTPS_PROXY_ADDRESS !== '')
+        {
+            if(!array_key_exists('http', $opts))
+            {
+                $opts['http'] = [];
+            }
+            $opts['http']['proxy'] = 'tcp://' . HTTPS_PROXY_ADDRESS . ':' . HTTPS_PROXY_PORT;
+            $opts['http']['request_fulluri'] = true;
+            if(HTTPS_PROXY_USERNAME !== '')
+            {
+                $headers = getValue($opts['http'], 'header', $defaultValue = []);
+                array_push($headers, 'Proxy-Authorization: Basic ' . base64_encode(HTTPS_PROXY_USERNAME . ':' . HTTPS_PROXY_PASSWORD));
+                $opts['http']['header'] = $headers;
+            }
+        }
         $context = getContextFromOpts($opts);
+        $url = 'https://checkip.amazonaws.com';
         $result = file_get_contents($url, false, $context);
+        die($result);
         return [$result, $http_response_header];
     }
 
diff --git a/configuration.php b/configuration.php
index b403e88..58ae206 100644
--- a/configuration.php
+++ b/configuration.php
@@ -7,6 +7,11 @@
     define('GOOGLE_ABUSE_EXEMPTION', '');
     define('MULTIPLE_IDS_ENABLED', True);
 
+    define('HTTPS_PROXY_ADDRESS', '');
+    define('HTTPS_PROXY_PORT', 80);
+    define('HTTPS_PROXY_USERNAME', '');
+    define('HTTPS_PROXY_PASSWORD', '');
+
     // No-key endpoint:
     define('KEYS_FILE', 'ytPrivate/keys.txt');
     // Both following entries can be generated with `tr -dc A-Za-z0-9 </dev/urandom | head -c 32 ; echo`.
```
</details>

**Comment by krasnoludkolo** [comment]
I'm not sure about `tcp://` prefix. Will it work with http(s) proxies?

**Comment by Benjamin-Loison** [comment]
The purpose of above `diff` is to work with HTTPS proxy, do not pay attention to `tcp` prefix.

**Comment by krasnoludkolo** [comment]
So I think it looks fine

**Comment by Benjamin-Loison** [comment]
Should not `HTTPS_PROXY_PORT` be *443*?

**Comment by Andromeda606** [comment]
I would like to be able to use multiple proxies and quickly switch between them. Like proxies.txt file

**Comment by Benjamin-Loison** [comment]
@Andromeda606 Why is not a single proxy meeting your needs?

**Comment by Andromeda606** [comment]
Proxies are usually provided more than once because they are likely to banned. 
If you have a residential proxy, you can request random requests from multiple proxies. Because the proxy can also be banned.

**Comment by Benjamin-Loison** [comment]
In #11 I have significantly studied workarounds to IP ban. As far as I remember the current state was whether YouTube was banning just the concerned IPv6 or also neighbor ones, if not, then could just easily change the IPv6 used. Have you tried that @Andromeda606?

**Comment by Andromeda606** [comment]
Nothing is currently banned. However, it could be, and taking precautions would be much better.

Changing the IP address on the server might be a difficult solution. Instead, a rotating proxy could be used, but multi-proxy support would be nice for this application

**Comment by Benjamin-Loison** [comment]
Similarly [blob/0d2768a5fcf560288eb3a9fa573056bdd5dba3d2/noKey/index.php](https://github.com/Benjamin-Loison/YouTube-operational-API/blob/0d2768a5fcf560288eb3a9fa573056bdd5dba3d2/noKey/index.php) has such rotation mechanism on quota exhaustion being detected.

You are free to share on this issue a `diff` showing the changes to support multiple proxies, but I won't merge it until I notice the issue and try an IPv6 workaround.

**Comment by Andromeda606** [comment]
I'm new to this API so I'm not sure if there will be any issues. I'll update this if there are any issues.
